### PR TITLE
Switch to PyPA build, installer, and flit-core to build Python packages

### PR DESCRIPTION
**Pull request by tjni**
I have updated the issue description with the latest status.

### Background

We are going to adopt  [`build`](https://github.com/pypa/build) as our bootstrap build front-end, [`flit-core`](https://github.com/pypa/flit) as our bootstrap build backend, and [`installer`](https://github.com/pypa/installer) as our bootstrap wheel installer.

This is based heavily on [this gist](https://gist.github.com/K900/6f8a56ece9240d3d101d3d03fc49e9d6) that @K900 shared with me. The only difference is that `build` is not used to bootstrap itself, which simplifies the build a little.

I want to acknowledge prior work from @FRidh in https://github.com/NixOS/nixpkgs/pull/105714 and @mweinelt in https://github.com/NixOS/nixpkgs/pull/232451, which introduced me to this idea.

### TLDR

There are a lot of commits in this PR – continue reading to find out why. I think this is a good approach in the long term, but I expect some breakages will be teased out on Hydra or afterward. I am confident that all packages can be fixed quickly once we've noticed a breakage.

### How do I review this monstrosity?

Thank you in advance!!! The description below will explain in more detail the changes, but I have split up the commits into the following sections:

1. The most important commits to review come first. Their descriptions should explain what they do. This includes the implementation of the bootstrapping, hooks, and key package updates. These are worth reviewing carefully.

2. What follows that is a series of commits that fix up packages to add build dependencies, patch version constraints, fix tests, etc. Each is small and self-explanatory. I suggest sampling some that seem interesting. I tried to use similar commit messages for similar changes, and they are ordered alphabetically by attribute path.

3. The final set of commits all have the commit message "add wheel dependency", and (unless I messed up by accident) they only add `wheel` to a package's nativeBuildInputs, because it is no longer being propagated automatically. I don't recommend reading these unless you really want to, but they are also ordered alphabetically by attribute path.

### Description of changes

This PR starts with a few commits updating maintainers, then about a dozen and a half of core commits, followed by a long, long tail of commits fixing things found based on the approach I am taking.

The core commits roughly do the following:

1. Create python3.pkgs.bootstrap.{build,flit-core,installer} packages.
4. Modify pypaBuildHook using build and introduce pypaInstallHook using installer.
5. Swap pipBuildHook for pypaBuildHook and pipInstallHook for pypaInstallHook.
6. Move tests for packages involved in bootstrapping to passthru to avoid cycles.
7. Update wheel to a version that builds with flit-core to break its dependency cycle with setuptools.
8. Clean up dependencies on boostrapped-pip.
9. Update setuptools and pip to the latest versions. This turned out not to be strictly necessary, but I had already tested a lot of things with it before I realized this.

The reason for this long tail of commits is based on a few differences / design decisions:

1. The old `pipBuildHook` adds `pip` and `wheel` to the PYTHONPATH. The new `pypaBuildHook` only includes `build`. This is a design decision: I think introducing as little into the environment as possible simplifies understanding how builds work. The consequence, however, is that any package that builds with `setuptools` implicitly needs `wheel` as defined in its PEP 517 [get_requires_for_build_wheel](https://peps.python.org/pep-0517/#get-requires-for-build-wheel) hook. When this is missing, build will fast-fail with an error message. I've done my best to identify the couple hundred of packages that need this added to its nativeBuildInputs, but I expect some will slip by and require fixing after the fact. Sorry!

2. Another consequence of this change is that `pip` is no longer part of the default PYTHONPATH, and so some packages need to add it explicitly in their tests.

3. `build` does both pyproject.toml validation and stricter dependency checking by default. This means some packages will need patching to its pyproject.toml or additional dependencies added to its nativeBuildInputs. I have also tried to identify as many of these as I could ahead of time.

4. The new version of setuptools issues a deprecation warning when `pkg_resources` is imported. Some projects turn all warnings into errors during tests and so will need to be patched to ignore the warning.

### Next steps

1. Appreciate any feedback on the approach. Even though there are a lot of commits here already, I don't care about sunk cost and don't mind backing out anything, doing it in smaller chunks, or anything else.

2. If this approach seems reasonable, then any feedback or help with testing is welcome. Note that these changes seem to be used in the build path of LLVM, so you will need to build some stuff if testing locally even if testing on a branch with Hydra artifacts. I will commit to keeping this PR up to date with staging when merge conflicts are discovered.

3. Documentation. I haven't gotten around to doing that yet.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by K900** [comment]
We shouldn't need a bootstrap pip/setuptools at all, we can build them as normal with this. Also, appreciate the work!

**Comment by FRidh** [comment]
Thanks for looking into this. It's badly needed and if am correct even blocking us on I think updating pip otherwise.

These are indeed the right steps. Comments inline.

> 1. Rebuild `flit-core` using bootstrapped `build` and `installer`.
>
> 2. Rebuild `build` using `flit-core`, bootstrapped `build`, and bootstrapped `installer`.
> 
> 3. Rebuild `installer` using `flit-core`, `build`, and bootstrapped `installer`.

From my point of view we can have a `bootstrapped-build ` and `bootstrapped-installer` that are then used in the hooks. The `bootstrapped-*` could be build by just installing all its deps in the same store path. Dumb and simple. Of course splitting things up further is fine but I think it should be optional.

> 4. Switch to [`pypaBuildHook`](https://github.com/NixOS/nixpkgs/pull/244135/commits/0f8f70af1115230387b3850b843350e56c154347) when pyproject is the format.
> 
> 5. Introduce and switch to `pypaInstallerHook` when pyproject is the format.
> 
> 6. Deprecate format = flit (?)

This can be done independently from this work. In my opinion it is OK to not even deprecate but outright remove the functionality. It was a temporary solution that was needed many years back. Nobody should depend on this since `pyproject` was introduced.

> 7. Clean up `bootstrapped-pip`.

We should be able to remove it.

>     8. Clean up `setuptools` bootstrapping.

**Comment by K900** [comment]
I think we should really have an extra step in the bootstrap to build `build` and `installer` properly, as they're supposed to also be used as modules by other tools (e.g. poetry) so we would need them as clean packages anyway.

**Comment by K900** [comment]
FWIW this entire thing will end up in staging anyway as soon as you touch any of the hooks and/or pip and/or setuptools, so I wouldn't worry about it too much. (re: last commit message)

**Comment by tjni** [comment]
I have rebased against staging and added a lot of commits. Don't be afraid, I updated the issue description to explain.

File: pkgs/development/interpreters/python/mk-python-derivation.nix [hunk 1]
```diff
@@ -184,15 +209,25 @@ let
       setuptoolsBuildHook
     ] ++ lib.optionals (format == "flit") [
       flitBuildHook
-    ] ++ lib.optionals (format == "pyproject") [
-      pipBuildHook
-    ] ++ lib.optionals (format == "wheel") [
+    ] ++ lib.optionals (format == "pyproject") [(
+      if python.isPy2 then
```
Comment (FRidh) [inline pkgs/development/interpreters/python/mk-python-derivation.nix:213]: The builder should not bother anymore with Python 2. For Python 2 there can be a copy of the builder, probably just the legacy one using pip.

File: pkgs/development/python-modules/bootstrap/build/default.nix [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+      runHook postBuild
+    '';
+
+    installPhase = ''
+      runHook preInstall
+
+      PYTHONPATH="${installer}/${python.sitePackages}" \
+          ${python.interpreter} -m installer \
+              --destdir "$out" --prefix "" dist/*.whl
```
Comment (FRidh) [inline pkgs/development/python-modules/bootstrap/build/default.nix:30]: What is the reason destdir is set here to $out instead of prefix?

**Review by FRidh — COMMENTED** [review]
Thanks for doing this, it looks great!

- Can you add the meta to the bootstrap packages as well?
- Is it possible to keep `bootstrapped-pip` and its hook around? I would not be surprised if somebody is using this directly. It would have to be deprecated.
- Have you tested with cross? I wonder whether that functions (with the nested bootstrap attribute set)

**Comment by FRidh** [comment]
> We should be able to remove it.

I now see I said it could be removed. Yes, we can remove it, but maybe not directly now.

File: pkgs/development/python-modules/bootstrap/build/default.nix (hunk 2 above)
Comment (tjni) [reply pkgs/development/python-modules/bootstrap/build/default.nix:30]: Great catch! This has been fixed in the latest version. I didn't know what the difference was, but it turns out there is an observable difference in the RECORD file, and using prefix is correct.

**Comment by K900** [comment]
Do we really need to explicitly add `wheel` everywhere? Maybe setuptools and friends should just propagate it.

**Comment by K900** [comment]
Also, if we do end up doing this, please squash the wheel-adding into one commit - the history is now absolutely impossible to browse on the web.

**Comment by tjni** [comment]
Thank you for your suggestions.

> Do we really need to explicitly add wheel everywhere? Maybe setuptools and friends should just propagate it.

I can see this perspective – it's good enough for all or nearly all cases, less work, and less breakage. My personal preference is to follow [PEP 517](https://peps.python.org/pep-0517/) exactly or choose the option that defines the dependency explicitly for clarity. To address the extra work and breakage issue, I'm willing to spend time perusing Hydra runs and bug reports to fix everything. I welcome counterpoints to this line of thought too.

> Also, if we do end up doing this, please squash the wheel-adding into one commit - the history is now absolutely impossible to browse on the web.

I will do this right now, didn't know about this GitHub limitation.

**Comment by K900** [comment]
Doesn't `build` inherently require `wheel` anyway?

**Comment by tjni** [comment]
It turns out that it does not. `build` delegates to a build backend, which could depend on `wheel`, such as setuptools, or it could reimplement its own logic for creating wheels, such as flit-core.

**Comment by K900** [comment]
If setuptools depends on wheel, then it should definitely just propagate it. We propagate all the Python dependencies because they need to be in the final environment to be importable, so this is perfectly normal and expected.

**Comment by FRidh** [comment]
> Do we really need to explicitly add `wheel` everywhere? Maybe setuptools and friends should just propagate it.

setuptools (pkg_resources) does not need wheel during runtime.


> I can see this perspective – it's good enough for all or nearly all cases, less work, and less breakage. My personal preference is to follow [PEP 517](https://peps.python.org/pep-0517/) exactly or choose the option that defines the dependency explicitly for clarity. To address the extra work and breakage issue, I'm willing to spend time perusing Hydra runs and bug reports to fix everything. I welcome counterpoints to this line of thought too.

I agree we should in Nixpkgs aim to be exact with our dependencies. Given the significant change in this PR, I can imagine a second follow-up PR to do this part correctly. Instead of adding wheel as runtime dependency, can we add it as build dependency by adding it to the build hook. Or does this cause trouble with bootstrapping?

> Doesn't `build` inherently require `wheel` anyway?

Actual wheel building is up to the backend itself, so that should be a no.

**Comment by tjni** [comment]
> setuptools (pkg_resources) does not need wheel during runtime.

Just to add onto this for clarity, because I didn't know this before I researched, PEP 517 introduces a way for a build backend to declare separate dependencies it needs when used as a build backend separately from when it is used as a library.

> Instead of adding wheel as runtime dependency, can we add it as build dependency by adding it to the build hook. Or does this cause trouble with bootstrapping?

Yeah, we can do it this way too. It does not cause any significant trouble (as far as I can tell). We could special case `wheel` like we do for build's dependencies (i.e. packaging and pyproject-hooks) so that it is built using the bootstrap build instead of the normal build. Everything else stays the same.

**Comment by tjni** [comment]
I'm going to move `wheel` into `pypaBuildHook` for now and move all of these fix up commits into a different PR (eventually). Aside from it being under discussion, I'm now realizing all these commits are hiding conversations, and I'd rather this be focused on the core changes.

**Comment by tjni** [comment]
I pushed a new version with only the commits that I think will still be needed. There is still some fix up I have to do to make this work, and I still need to address feedback, which I'll do and test tomorrow.

The changes are:

1. Add `wheel` to propagatedBuildInputs of `pypaBuildHook`, letting us drop all of the commits that only add wheel.
2. Add `--skip-dependency-check` to `pypaBuildHook`, letting us drop the commits that fix up build constraints.

We still need to keep commits that:

1. Add pip as a test dependency, since it is no longer available from the build hook.
2. Fix builds that need patches to be compatible with the new versions of pip or setuptools.
3. Patches that fix malformed pyproject.toml files.

**Comment by FRidh** [comment]
As soon as you think the majority builds for you as far as you can test, we can run it on hydra as well. Best check with @mweinelt and @SuperSandro2000 regarding availability of hydra.

**Comment by mweinelt** [comment]
I think we want to a dedicated jobset for python-bootstrap in that scenario, so we can continue using python-updates in parallel. WDYT?

In that case we'd best ask on the infra room for someone to create it.

**Comment by FRidh** [comment]
At d8bb14539adc6efd2a0487624ef686f890d9080c I get
```
$ nix-build -A python3.pkgs.xarray
error: infinite recursion encountered

       at /home/freddy/code/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:344:7:

          343|       depsBuildBuild              = lib.elemAt (lib.elemAt dependencies 0) 0;
          344|       nativeBuildInputs           = lib.elemAt (lib.elemAt dependencies 0) 1;
             |       ^
          345|       depsBuildTarget             = lib.elemAt (lib.elemAt dependencies 0) 2;
(use '--show-trace' to show detailed location information)
```

**Comment by tjni** [comment]
Thank you for testing this! It's because the latest commit I pushed added wheel to pypaBuildHook but forgot to exclude it when building wheel itself. I'm going to fix this.

This aside, I'm running into a separate problem and can use some advice on how to proceed. I added `--skip-dependency-check` when running build in the hope that it will reduce the number of packages that need build dependency patches, but to my surprise it broke `setuptools` own build.

I tracked it down to https://github.com/pypa/setuptools/issues/2586, which suggests running the `egg_info` command before building it with a PEP 517 backend, but why does it pass without `--skip-dependency-check`? I suspect it's because setuptools [calls egg_info internally](https://github.com/pypa/setuptools/blob/80e60c3bdd5aa57f32560e5b0d946d0c07c71ac2/setuptools/build_meta.py#L316) when it is queried for its build dependencies.

> I think we want to a dedicated jobset for python-bootstrap in that scenario, so we can continue using python-updates in parallel. WDYT?

I was initially thinking to reuse `python-updates` during a calm period (maybe now?) and do a "release" of it when these changes (in whichever form we decide) are validated, and pick up any dependency updates we want to do along the way.

From my observation, there are only a small handful of failure conditions, all which can be worked around by patching. Even with the aggressive approach of dropping wheel and keeping the new build validations, I was able to build and fix about 600 or so issues locally in three days. I have loads of time right now, so with some help committing, reviewing, and scheduling jobs, I ambitiously think this validation could be done in two weeks.

**Comment by mweinelt** [comment]
My plan was to start defaulting to python3.11 in the coming week.

**Comment by tjni** [comment]
I just pushed a minor change to fix the infinite recursion (which I think is related to `wheel`). It does not include the `--skip-dependency-checks` part that causes problems. Please feel free to play around with it if you guys want, and thank you for doing so.

**Comment by tjni** [comment]
Feedback response:

> The builder should not bother anymore with Python 2. For Python 2 there can be a copy of the builder, probably just the legacy one using pip.

I created a copy of mk-python-derivation in a commit for Python 2. The only changes from the original one are:

1. It removes support for format == "flit", since that doesn't work on Python 3.
2. It removes usage of `pythonNamespacesHook`, which was only used for Python >= 3.3.

> Can you add the meta to the bootstrap packages as well?

The bootstrap packages currently inherit meta from the main packages under `python-modules`. I hope that's OK.

> Is it possible to keep bootstrapped-pip and its hook around? I would not be surprised if somebody is using this directly. It would have to be deprecated.

I dropped the commit removing it.

> Have you tested with cross? I wonder whether that functions (with the nested bootstrap attribute set)

I have not, I'll figure out how to test it.

---

With help from @adisbladis, I added a commit that updates poetry2nix's wheel override so that it should work with the new wheel derivation.

File: pkgs/top-level/python-packages.nix [hunk 3]
```diff
@@ -8,6 +8,16 @@
...
+    flit-core = toPythonModule (callPackage ../development/python-modules/bootstrap/flit-core { });
+    installer = toPythonModule (callPackage ../development/python-modules/bootstrap/installer {
+      inherit (bootstrap) flit-core;
+    });
+    build = toPythonModule (callPackage ../development/python-modules/bootstrap/build {
+      inherit (bootstrap) flit-core installer;
+    });
+  };
+
   bootstrapped-pip = toPythonModule (callPackage ../development/python-modules/bootstrapped-pip { });
```
Comment (SuperSandro2000) [inline pkgs/top-level/python-packages.nix:21]: Not sure if we should move this also into bootstrap

File: pkgs/development/python-modules/installer/default.nix [hunk 4]
```diff
@@ -13,24 +14,43 @@ buildPythonPackage rec {
...
-    homepage = "https://github.com/pradyunsg/installer";
     description = "A low-level library for installing a Python package from a wheel distribution";
+    homepage = "https://github.com/pypa/installer";
+    changelog = "https://github.com/pypa/installer/blob/${src.rev}/docs/changelog.md";
     license = licenses.mit;
-    maintainers = with maintainers; [ cpcloud fridh ];
+    maintainers = teams.python.members ++ [
+      maintainers.cpcloud
+      maintainers.tjni
+    ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/installer/default.nix:54]:
```suggestion
    maintainers = teams.python.members ++ (with maintainers; [ cpcloud tjni ];
```

File: pkgs/development/python-modules/cypari2/default.nix [hunk 5]
```diff
@@ -37,7 +37,7 @@ buildPythonPackage rec {
     export PYTHONPATH="$out/${python.sitePackages}:$PYTHONPATH"
 
     # install "." instead of "*.whl"
-    ${python.pythonForBuild.pkgs.bootstrapped-pip}/bin/pip install . --no-index --no-warn-script-location --prefix="$out" --no-cache
+    ${python.pythonForBuild.pkgs.pip}/bin/pip install . --no-index --no-warn-script-location --prefix="$out" --no-cache
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/cypari2/default.nix:40]: We should add pip to nativeBuildInputs instead if possible

File: pkgs/development/python-modules/colour/default.nix [hunk 6]
```diff
@@ -1,15 +1,39 @@
...
+  ];
+
+  nativeCheckInputs = [
+    nose3
+  ];
+
+  checkPhase = ''
+    runHook preCheck
+
+    nosetests -sx .
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/colour/default.nix:29]: Running the tests with pytest does not just work?

File: pkgs/development/python-modules/build/default.nix [hunk 7]
```diff
@@ -86,6 +103,9 @@ buildPythonPackage rec {
     homepage = "https://github.com/pypa/build";
     changelog = "https://github.com/pypa/build/blob/${version}/CHANGELOG.rst";
     license = licenses.mit;
-    maintainers = with maintainers; [ fab ];
+    maintainers = teams.python.members ++ [
+      maintainers.fab
+      maintainers.tjni
+    ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/build/default.nix:109]:
```suggestion
    maintainers = teams.python.members ++ (with maintainers; [ fab tjni ];
```

File: pkgs/development/python-modules/bootstrap/installer/default.nix [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        ${python.interpreter} -m flit_core.wheel
+
+    runHook postBuild
+  '';
+
+  installPhase = ''
+    runHook preInstall
+
+    PYTHONPATH=src ${python.interpreter} -m installer \
+        --destdir "$out" --prefix "" dist/installer-*.whl
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/bootstrap/installer/default.nix:25]:
```suggestion
      --destdir "$out" --prefix "" dist/installer-*.whl
```

File: pkgs/development/python-modules/bootstrap/installer/default.nix [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+stdenv.mkDerivation {
+  pname = "${python.libPrefix}-bootstrap-${installer.pname}";
+  inherit (installer) version src meta;
+
+  buildPhase = ''
+    runHook preBuild
+
+    PYTHONPATH="${flit-core}/${python.sitePackages}" \
+        ${python.interpreter} -m flit_core.wheel
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/bootstrap/installer/default.nix:16]:
```suggestion
     ${python.interpreter} -m flit_core.wheel
```

File: pkgs/development/python-modules/bootstrap/flit-core/default.nix [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    ${python.interpreter} -m flit_core.wheel
+
+    runHook postBuild
+  '';
+
+  installPhase = ''
+    runHook preInstall
+
+    ${python.interpreter} bootstrap_install.py dist/flit_core-*.whl \
+        --install-root "$out" --installdir "/${python.sitePackages}"
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/bootstrap/flit-core/default.nix:25]:
```suggestion
      --install-root "$out" --installdir "/${python.sitePackages}"
```

File: pkgs/development/python-modules/bootstrap/build/default.nix [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+let
+  buildBootstrapPythonModule = basePackage: attrs: stdenv.mkDerivation ({
+    pname = "${python.libPrefix}-bootstrap-${basePackage.pname}";
+    inherit (basePackage) version src meta;
+
+    buildPhase = ''
+      runHook preBuild
+
+      PYTHONPATH="${flit-core}/${python.sitePackages}" \
+          ${python.interpreter} -m flit_core.wheel
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/bootstrap/build/default.nix:20]:
```suggestion
        ${python.interpreter} -m flit_core.wheel
```

File: pkgs/development/python-modules/bootstrap/build/default.nix (hunk 2 above)
Comment (SuperSandro2000) [inline pkgs/development/python-modules/bootstrap/build/default.nix:30]:
```suggestion
      PYTHONPATH="${installer}/${python.sitePackages}" \
        ${python.interpreter} -m installer \
          --destdir "$out" --prefix "" dist/*.whl
```

File: pkgs/development/interpreters/python/mk-python2-derivation.nix [hunk 12]
```diff
@@ -0,0 +1,252 @@
+# Generic builder.
```
Comment (SuperSandro2000) [inline pkgs/development/interpreters/python/mk-python2-derivation.nix:1]:
```suggestion
# Generic builder only used for EOL and deprecated python2
```
lets make it clear that people should not continue to work on this

**Review by SuperSandro2000 — COMMENTED** [review]
Not so super happy of having the 4th or so variant of running tests in passthru. We should unify that at some point.

File: pkgs/top-level/python-aliases.nix [hunk 13]
```diff
@@ -77,6 +77,7 @@ mapAliases ({
   coronavirus = throw "coronavirus was removed, because the source is not providing the data anymore."; # added 2023-05-04
   cozy = throw "cozy was removed because it was not actually https://pypi.org/project/Cozy/."; # added 2022-01-14
   cryptography_vectors = "cryptography_vectors is no longer exposed in python*Packages because it is used for testing cryptography only."; # Added 2022-03-23
+  d2to1 = throw "d2to1 is archived and no longer works with setuptools v68"; # added 2023-07-30
```
[the remark this answers is not in the document]
Comment (mweinelt) [reply pkgs/top-level/python-aliases.nix:80]: Does that really matter much?

[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply pkgs/top-level/python-aliases.nix:80]: no, not really but it would be something to reduce the size of this PR making it easier to review

File: pkgs/development/python-modules/colour/default.nix (hunk 6 above)
Comment (tjni) [reply pkgs/development/python-modules/colour/default.nix:29]: Thank you for the poke. It didn't when I tried initially, but I took a closer look and found it just needs some flags.

File: pkgs/top-level/python-aliases.nix (hunk 13 above)
[the remark this answers is not in the document]
Comment (tjni) [reply pkgs/top-level/python-aliases.nix:80]: I agree with both of you, and won't turn down an opportunity to merge part of this faster :) I opened https://github.com/NixOS/nixpkgs/pull/246829.

File: pkgs/development/python-modules/cypari2/default.nix (hunk 5 above)
Comment (tjni) [reply pkgs/development/python-modules/cypari2/default.nix:40]: Good idea, it seems to work!

File: pkgs/top-level/python-packages.nix (hunk 3 above)
Comment (tjni) [reply pkgs/top-level/python-packages.nix:21]: I like this for consistency, but I am hoping instead to leave this here and pursue deprecating this for Python 3. I just need to think of a good way to do it (docs, warnings, something else). Opinions appreciated.

**Comment by tjni** [comment]
Thank you for your reviews. I have addressed all of the feedback so far, and it's now ready to be tested on Hydra. Callouts:

1. I have renamed the new `mk-python2-derivation.nix` to `python2/mk-python-derivation.nix` for better organization. I see some other opportunities potentially to improve some infrastructure for Python 3 in the future that still requires keeping around artifacts for Python 2.

2. I added a commit to no longer propagate `wheel` from `pythonRelaxDepsHook`. It's being propagated in `pypaBuildHook` for now already and might be removed in the future, and I don't want this hook to complicate that.

3. The hint to test cross compilation was a great one. Now that I switched `pypaBuildHook` to use the `pyproject-build` CLI, it needs to be built using `python.pythonForBuild`. For consistency, I also used `installer` from pythonForBuild too even though that isn't strictly necessary.

4. I think it is a good idea to start extracting out some of the follow up commits into their own PRs to make this one easier to review if they don't strictly depend on this one. I'm going to start doing that so we have the option.

**Comment by tjni** [comment]
I forgot to add documentation, I'll work on that too.

File: pkgs/development/python-modules/cypari2/default.nix [hunk 14]
```diff
@@ -37,11 +37,12 @@ buildPythonPackage rec {
     export PYTHONPATH="$out/${python.sitePackages}:$PYTHONPATH"
 
     # install "." instead of "*.whl"
-    ${python.pythonForBuild.pkgs.bootstrapped-pip}/bin/pip install . --no-index --no-warn-script-location --prefix="$out" --no-cache
+    pip install . --no-index --no-warn-script-location --prefix="$out" --no-cache
   '';
 
   nativeBuildInputs = [
     pari
+    python.pythonForBuild.pkgs.pip
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/cypari2/default.nix:45]: Naive question: Just adding pip is not splicing correctly?

Comment (tjni) [reply pkgs/development/python-modules/cypari2/default.nix:45]: You probably know more about this than I do. I don't yet know how the splicing is implemented, and I was not able to build up to `cypari2` with the cross build command I used. For now I left this as-is and only tested building cypari2 non-cross to limit the risk, but this is something I'd like to look into when I have time.

**Comment by tjni** [comment]
There were three build-time package dependencies that quite a number of projects had which needed patching if we are to keep the build-time dependency validation: [cmake](https://github.com/scikit-build/cmake-python-distributions), [ninja](https://github.com/scikit-build/ninja-python-distributions), and [oldest-supported-numpy](https://github.com/scipy/oldest-supported-numpy).

I thought about this a bit and now feel that a good way forward is to stub out these three:

1. https://github.com/NixOS/nixpkgs/pull/246963
2. https://github.com/NixOS/nixpkgs/pull/247064
3. https://github.com/NixOS/nixpkgs/pull/247073

For cmake and ninja, I feel that shadowing the top-level names is reasonable. Right now, many projects that require these tools pick them up from the PyPI packages, and creating these drop-in replacements means they now satisfy dependency validation without any work.

On the other hand, there are not many projects that can depend on the tools directly (meson-python, scikit-build-core, and any projects that use scikit-build-core and have removed their cmake PyPI dependency, of which I could only find one so far in nixpkgs). For these types of projects, even if we forget to depend on the tool directly, depending on the PyPI package should still work.

To everyone: please feel free to share any thoughts or opinions you have too.

File: pkgs/development/python-modules/cypari2/default.nix (hunk 14 above)
Comment (FRidh) [reply pkgs/development/python-modules/cypari2/default.nix:45]: The splicing will need to work. If it doesn't yet then we have a regression and it would be likely that the issue occurs in more cases.

**Comment by tjni** [comment]
Helpfully, the build for [asn2quickder.x86_64-linux](https://hydra.nixos.org/build/231289481/nixlog/1) yielded an error:
```
Executing pypaInstallPhase
usage: python -m installer
       [-h]
       [--destdir path]
       [--prefix path]
       [--compile-bytecode level]
       [--no-compile-bytecode]
       wheel
python -m installer: error: unrecognized arguments: dist/arpa2.quickder_tools-1.7.1-py3-none-any.whl
```
This was because we did not handle installing when a build creates more than one wheel. I updated pypaInstallHook to do this properly now in a loop.

**Comment by FRidh** [comment]
> Upstream's recommended "python -m build" way of invoking build fails
when the working directory contains a file named "build.py". This is
common for poetry projects that build C extensions.

Has this been reported upstream (PyPA/build)? It's inevitable that name clashes can occur, but by choosing a different name it does get less likely.

I am testing this now in hydra as well https://hydra.nixos.org/jobset/nixpkgs/python-updates. Please avoid force pushing. I took your commits and pushed them to the `python-updates` branch.

**Comment by FRidh** [comment]
> This was because we did not handle installing when a build creates more than one wheel. I updated pypaInstallHook to do this properly now in a loop.

It's better to install them simultaneously in a single invocation. Python dependencies can be circular, they do not need to be on a DAG. Hence, it can be possible the only way to install packages is to install them together.

**Comment by tjni** [comment]
> Has this been reported upstream (PyPA/build)? It's inevitable that name clashes can occur, but by choosing a different name it does get less likely.

The closest issue I could find to this is https://github.com/pypa/build/issues/526, in which a contributor acknowledges that the naming is unfortunate. I think it's too late to change, and luckily (or by design) they ship a CLI named pyproject-build that we can use instead.

> Please avoid force pushing. I took your commits and pushed them to the `python-updates` branch.

Thank you for kicking off the test run, I've been using its results today. I'll make sure any changes I make are new commits. 

> It's better to install them simultaneously in a single invocation.

I didn't see a way in the documentation of [installer](https://installer.pypa.io/en/stable/cli/installer/) to do this or an issue in their repository when I did a quick search. Does installation need to happen in a particular order? I was under the impression that this step doesn't do any dependency checking.

**Comment by FRidh** [comment]
I just started a new evaluation on hydra with this change rebased onto staging. https://hydra.nixos.org/jobset/nixpkgs/python-updates

File: pkgs/development/python-modules/pip/default.nix [hunk 15]
```diff
@@ -15,14 +15,14 @@
 
 buildPythonPackage rec {
   pname = "pip";
-  version = "23.0.1";
+  version = "23.2.1";
```
Comment (doronbehar) [inline pkgs/development/python-modules/pip/default.nix:18]: I have a small request: [This upstream fix](https://github.com/pypa/pip/pull/11417/commits/fb454572b2e455b22a59822b53b203b7232570e6) includes `pip completion` fixes I've been waiting for a long time. Could we add those completion files via `installShellCompletion`?

**Review by doronbehar — COMMENTED** [review]
Also, ofborg evaluation fails currently.

File: pkgs/development/python-modules/pip/default.nix (hunk 15 above)
Comment (tjni) [reply pkgs/development/python-modules/pip/default.nix:19]: Of course, I added a commit that should do this. I'm not sure how long this whole thing will take to finish, however.

**Comment by tjni** [comment]
> Also, ofborg evaluation fails currently.

Thanks for bringing it to my attention. I'm avoiding rebasing right now to not mess up the testing on python-updates that @FRidh is helping me with, but it will be resolved prior to merging.

File: pkgs/development/python-modules/pip/default.nix (hunk 15 above)
Comment (doronbehar) [reply pkgs/development/python-modules/pip/default.nix:19]: > Of course, I added a commit that should do this. I'm not sure how long this whole thing will take to finish, however.

Sure. No rush about it :). Thanks for all your work!

**Comment by FRidh** [comment]
> I just started a new evaluation on hydra with this change rebased onto staging. https://hydra.nixos.org/jobset/nixpkgs/python-updates

Hydra seems to be occupied. Looks like we need to do it without it.

**Comment by tjni** [comment]
That's ok, still plenty to do, and the last run yielded some good information. When the staging-next queue clears up, we might be able to get a few more builds in (with hopefully some more fixes too).

**Comment by FRidh** [comment]
@tjni could you add further changes to #248866.

**Comment by tjni** [comment]
Ah, I can but I think I need to get write permissions first on the repo to push to that PR. I'll send in a request tomorrow.

**Comment by doronbehar** [comment]
This can be closed right?

**Comment by tjni** [comment]
Closing in favor of https://github.com/NixOS/nixpkgs/pull/248866.
